### ISSUE TITLE
Static port clearance for service provisioning?

### ISSUE THREAD

**Issue by NetDevils** [gerbera/gerbera#1512]
I am running Gerbera on my Raspberry Pi with Ubuntu 20.04. I have an ufw activated and can access the Gerbera WebUI, however my Android phone's "BubbleUPnP" app is not able to find Gerbera service and list the contents I share. Not as long as I deactivate ufw and restart Gerbera. It then shows it the service immediately in my app.

Using `sudo lsof -i -P -n | grep gebera` I see:
- 3 entries on port 52003/tcp (which is my WebUI port) and is whitelisted in ufw
- 3 entries on port 1900/udp (which is the streaming port I guess?) and is whitelisted in ufw
- 1 entry on an everychanging port (udp) which I guess is the broadcast for finding the service? (ports > 40.000, e.g. 40831 or 52631 or 54519)

Is there a way to make that static thus I can whitelist it?

By coincidence I have just checked and somehow it seems to working now? It was not working for months... ufw is active, whiteslisted are only the 6 ports above without the "random one" and I am still able to access the service from my smartphone. Even after several Gerbera.service restarts

Edit: Nope, 5 minutes later and the service is gone from BubbleUPnP ... maybe it was just cached or something.

**Comment by whyman** [comment]
1900/udp and outbound specifically the broadcast address 239.255.255.250 and the TCP port for the web-ui and content is all that should be needed. The other port is likely us discovering devices on the network, but isnt strictly needed.

**Comment by NetDevils** [comment]
```
me@myserver:/usr/share/gerbera/web$ sudo ufw status numbered
Status: active

     To                         Action      From
     --                         ------      ----
[ 6] 1900                       ALLOW IN    Anywhere
[ 7] 52003                      ALLOW IN    Anywhere
[15] 1900 (v6)                  ALLOW IN    Anywhere (v6)
[16] 52003 (v6)                 ALLOW IN    Anywhere (v6)
```

Those are cleared within ufw (TCP and UDP both).
ufw is by default set to allow all outgoing traffic.

But at the moment neither my Windows PC nor my Android App can detect the server.
Can you recommand any apps on Windows / Android for troubleshooting? Just to rule them out.

Not sure about the broadcast you mentioned.

And thanks for the quick replies! :-)

**Comment by KarlStraussberger** [comment]
In order for broadcasts to work you need to follow the steps described under network https://docs.gerbera.io/en/stable/run.html#network-setup

**Comment by NetDevils** [comment]
Hi Karl,
that did unfortunately not work...

First of all it did not work and second of all the routes got deleted upon restart.
Pi is running on LAN, thus eth0 

```
route add -net 239.0.0.0 netmask 255.0.0.0 eth0
ifconfig eth0 allmulti
```
===> not found in BubbleUPnP (which is not the renderer in the end. The renderer is going to be a Chromecast, but BubbleUPnP will be used for controlling the renderer. As said it works as long as ufw is disabled.)

I thought I might need to reboot. Did so: rules were gone...

I then added both commands to /etc/init.d/gerbera and rebooted, but the routes did not get included in the table at all. 
The docs.gerbera article mentions `Those settings will be applied automatically by the init.d startup script.`. Really? How?

Do I need to change any of the parameters according to my DHCP IP ranges used? (192.168.1.xxxxx by my AVM Fritzbox Router which is the DHCP server. Grüße an Karl nach Deutschland :-) [Greetings to fellow German Karl] )

Also man page of route says it is outdated an "ip route" should be used. Might that be an issue?

Sorry to bother you with that many questions, but although I am technically familiar this level of Unix know how is yet way beyond my limits.

Cheers and thanks for your help!

**Comment by KarlStraussberger** [comment]
The startup script is the second best place to add the commands. It may not be used by your raspi at all.

Caution: Gerbera must not be started before the network.

I simply use the systemd unit file for startup and have the broadcast config in network files. 

You can translate the `route` options to `ip route` to be future proof but the commands still work without new features.

You need to make sure that the broadcast is up before gerbera sends the first ssdp message (udp 1900). The best way is to test everything manually and the stabilize the boot process - only needed upon kernel update.

**Comment by KarlStraussberger** [comment]
Ach ja - Danke
_by the way - thanks_

**Comment by NetDevils** [comment]
Going down the rabbit hole, I see.

> The best way is to test everything manually and the stabilize the boot process 

Well that did not work obviously, did it? I did the "sudo route" and "sudo ifconfig" commands manually and "sudo systemctl restart gerbera.service", but Gerbera was not listed in my BubbleUPnP app. Or was this already too late? Or is there a command to restart Gerbera on a deeper level?

> have the broadcast config in network files

in /etc/rc2.d/ is the first Gerbera start script as far as I can see ("/etc/rc2.d/S01gerbera -> ../init.d/gerbera"). In rc0.d and rc1.d there is one file each K01gerbera. Going in lexical sort I just need to place a symbolic link /etc/rc2.d/S01gerbePreStart to a script file that executes the route and ifconfig commands?

On the other hand I should make sure that the script is executed by Ubuntu on startup anyways. Maybe just a small logging into a text file or so.
As said I really appreciate your help as this is getting way deeper than I am familiar with. But in the end I would also love to have a firewall running 24/7 on my homeserver and not having to manually disable it all the time just to be able to watch a movie :-)

Bye

**Comment by KarlStraussberger** [comment]
If you still have active startup script, you can add your lines there. No need to create a separate file.

Just to make sure, try everything with your firewall down. I'm not sure whether it blocks broadcasts by default. Maybe you need to open  the broadcast address also.

**Comment by KarlStraussberger** [comment]
No worries, once we're finished, this topic may go into some docs...

**Comment by NetDevils** [comment]
All righty. Managed to symlink a script into rc2.d which adds above mentioned route and ifconfig calls before Gerbera is started (named it S00...... in contrast to Gerbera's S01.... file).

Route is available after restart:
```
me@server:~$ sudo route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
239.0.0.0       0.0.0.0         255.0.0.0       U     0      0        0 eth0
```
Same for all multicast
```
netdevil@homeserver-pi:~$ sudo ifconfig
eth0: flags=4675<UP,BROADCAST,RUNNING,ALLMULTI,MULTICAST>  mtu 1500
        inet 192.168.1.20  netmask 255.255.255.0  broadcast 192.168.1.255
```

- If I **enable ufw and restart** the Pi I am not able to find Gerbera on my smartphone / BubbleUPnP.
- Once I **disable ufw** I am nearly immediately able to find Gerbera on my smartphone / BubbleUPnP. Even without restart Pi or restarting Gerbera service.

I enabled logging in ufw (setting to highest) and in the ufw logs I have:

- 457 entries created during the test case
- - only 33 of them are with status "**BLOCKED**". And of those 33 there are
- - - 6 from my Router
- - - 7 from one of the Smartphones (running BubbleUPnP) and 15 from another Smartphone (running BubbleUPnP).
- - - Only 1 of them is non-TCP (all other TCP). The non-TCP is: `[   86.103895] [UFW BLOCK] IN=eth0 OUT= MAC=01:00:5e:00:00:01:44:4e:6d:ef:76:bc:08:00 SRC=192.168.1.10 DST=224.0.0.1 LEN=36 TOS=0x00 PREC=0xC0 TTL=1 ID=0 DF PROTO=2 `
(192.168.1.10 being my Router)

On the other hand I do have 12 hits in the logs with connections being **allowed** from the Pi to the broadcast IP:
`[UFW ALLOW] IN= OUT=eth0 SRC=192.168.1.20 DST=239.255.255.250 LEN=470 TOS=0x00 PREC=0x00 TTL=4 ID=10994 DF PROTO=UDP SPT=34333 DPT=1900 LEN=450 `

Bottom line: I still have no clue why this is happening... ufw is definitely responsible for blocking the connection, but I am not sure why.

**Comment by KarlStraussberger** [comment]
It may be the incoming ssdp search request from bubble that's block. Udp/1900 has to be open in and out.

**Comment by NetDevils** [comment]
That's the thing, those are already set...

```
my@server:~$ sudo ufw status verbose
Status: active
Logging: off
Default: deny (incoming), allow (outgoing), disabled (routed)
```

Outgoing is allowed to everything
Incoming is allowed only to the specified rules. 

Compare above:
```
me@myserver:/usr/share/gerbera/web$ sudo ufw status numbered
Status: active

     To                         Action      From
     --                         ------      ----
[ 6] 1900                       ALLOW IN    Anywhere
[15] 1900 (v6)                  ALLOW IN    Anywhere (v6)
```
(no protocol display is valid for both TCP and UPD)

**Comment by NetDevils** [comment]
Update: this really does not seem to be a Gerbera issue alone. Just installed tvheadend and the identification of the available DVB-C interfaces on my network is done via SSDP as well. Had to deavtivate ufw for tvheadend to find the adapters. Just another reason for looking into the topic.

**Comment by KarlStraussberger** [comment]
Just another point. The incoming ssdp search is on `239.255.255.250`

**Comment by NetDevils** [comment]
Still not working :-/
I added an "ALLOW from 239.255.255.250 to ANYWHERE" as well as "ALLOW from 244.0.0.1 to ANYWHERE" and still the broadcasts are being rejected ... Digging through the internet I found few threads having the same issues. Basically it seems ufw needs to be foreced to allow those packages.
Them here [https://ubuntuforums.org/showthread.php?t=2445937](url) are talking about a very similar issue but I am not sure where they add their `-p UDP --sport 1900 -s 192.168.0.0/16 rule` solution. Iptables / ufw rules via CLI or ufw rules in some config...

I will go and try to find an ufw / Ubuntu expert board and discuss with them. Afterwards I will come back to share the solution (hopefully!)

**Comment by NetDevils** [comment]
Just an update so you know I haven' ditched the topic.
It definitly seems to be a problem with ufw and not with Gerbera or Tvheadend, thus you are totally off the hook.

At the moment I kind of managed to have ufw active and Gerbera + Tvheadend running. It all comes down to the UPnP / SSDP packages and open ports needed. It looks like TvHeadend requires UDP incoming ports > 52000 to be opened and randomly selects any of them for connection.
For Gerbera it is the same fixed one for the WebUI as for content delivery I guess. 

So at the moment I have sh**load of ports open and will need to strip that down casually over the time. The impacts are not visible directly but rather after a couple of minutes / restarts / hours only (?!?!?).

The most promising ones are:
FROM 239.255.255.250/udp to ANYWHERE locally
FROM ANYWHERE to 1900/udp locally

as well as this fancy piece of code I put in my init script to run before init.d/gerbera is loaded:
```
ipset create upnp hash:ip,port timeout 10
iptables -A OUTPUT -d 239.255.255.250/32 -p udp -m udp --dport 1900 -j SET --add-set upnp src,src --exist
iptables -A INPUT -p udp -m set --match-set upnp dst,dst -j ACCEPT
```
(Source [https://serverfault.com/a/911286](url))

**Comment by whyman** [comment]
Looks like we can close this one :+1:

**Comment by NetDevils** [comment]
Yeah sorry ... Haven't found the time to really trial and error me through the different scenarios :-/
Bottom line it is a configuration issue with ufw on my PC and will not affect to many people in the future. And even if there is silver lining posted above to mitigate the problem a bit
